woocommerce/woocommerce-android
# Add unread product reviews filter

Add unread product reviews filter


### Description

As part of the Hack Week project, I added a new filter for showing only `unread` product reviews. This comes from some user feedback stating things like: 

> Hi tech team. I suggest you change the reviews, so it’s possible to see which reviews the user hasn’t read. I have 2 unread reviews and I have opened all current reviews and now I need to scroll to the bottom and reload older reviews. Also when returning from a specific review, return to where you were in the list of reviews. Thank you.

How filtering unread product reviews work: 
- The `unread` state of product reviews relies on WPcom notifications `unread` status. 
- WC Rest API doesn't provide a way to fetch unread product reviews, as it doesn't keep track of this status. So we have to deal with this locally. 
- In the apps what we do is retrieve all the `unread` product review notifications from DB. Then we use `meta?.ids?.comment` field from the notification model to get the product review id. 
- With the list of unread product review ids we can now fetch from the API those specific reviews by setting `productIds: List<Long>?` field inside the `FetchProductReviewsPayload` 
- The rest of the code changes are essentially to handle pagination properly for unread product reviews. The unread reviews will be loaded in chunks of 25 items until all unread reviews are fetched. 

Additionally, this PR gets rid of some code duplication and removes `https://github.com/woocommerce/woocommerce-android/pull/9753/files#diff-67759c42aa62e65e7e1f76c809cd9804580b532574f8ee909481f2db6cecf0c2` from the code and refactors `ReviewListRepository` to support fetching product reviews for a specific product.  


### To Improve
When caching product reviews there are some scenarios when reopening product reviews list will display an outdated list for a couple seconds (until the new data is loaded). This is specially noticeable when you:
- Open product reviews
- Apply unread filter
- Close product reviews
- Reopen it and see how only unread reviews are shown for a couple seconds. 


### Testing instructions

1. Open a store that has a considerable amount of product reviews. You can [use this plugin to generate fake product reviews](https://wordpress.org/plugins/woo-virtual-reviews/). Which by the way I don't know how that is even allowed. Pretty sketchy 😅 
2. Mark several reviews as read
3. Enable and disable the unread filter and check the list is updated correctly. 

### Images/gif


**Unread reviews filter demo**

https://github.com/woocommerce/woocommerce-android/assets/2663464/82965416-c1ab-4487-8078-925030ce93a4

**Empty unread reviews UI**

****

https://github.com/woocommerce/woocommerce-android/assets/2663464/cb5ac5b6-dc0f-4be5-a88d-b2ff5de86fc0
